mutiny-core/src/logging.rs

```diff
@@ -1,9 +1,90 @@
+use std::sync::{
+    atomic::{AtomicBool, Ordering},
+    Arc, Mutex,
+};
+
+use crate::{error::MutinyError, utils::sleep};
 use chrono::Utc;
+use gloo_utils::format::JsValueSerdeExt;
 use lightning::util::logger::{Level, Logger, Record};
 use log::*;
+use rexie::{ObjectStore, Rexie, TransactionMode};
+use wasm_bindgen::JsValue;
+use wasm_bindgen_futures::spawn_local;
+
+pub(crate) const LOGGING_DATABASE_NAME: &str = "logging";
+pub(crate) const LOGGING_OBJECT_STORE_NAME: &str = "log_store";
+pub(crate) const LOGGING_KEY: &str = "logs";
+
+#[derive(Default, Clone)]
+pub struct MutinyLogger {
+    should_write: bool,
+    memory_logs: Arc<Mutex<Vec<String>>>,
+}
+
+impl MutinyLogger {
+    pub(crate) fn with_writer(stop: Arc<AtomicBool>) -> Self {
+        let l = MutinyLogger {
+            should_write: true,
+            memory_logs: Arc::new(Mutex::new(vec![])),
+        };
+
+        let log_copy = l.clone();
+        spawn_local(async move {
+            let logging_db = build_logging_database().await;
+            if logging_db.is_err() {
+                error!("could not build logging database, log entries will be lost");
+                return;
+            }
+            let logging_db = logging_db.unwrap();
+
+            loop {
+                // wait up to 5s, checking graceful shutdown check each 1s.
+                for _ in 0..5 {
+                    if stop.load(Ordering::Relaxed) {
+                        logging_db.close();
+                        return;
+                    }
+                    sleep(1_000).await;
+                }
```

Comment:
I'll probably go back and do this everywhere in a follow up PR. Slims down on the stops taking too long.